packages/next/src/server/image-optimizer.ts

```diff
@@ -438,6 +451,21 @@ export function getMaxAge(str: string | null | undefined): number {
   return 0
 }
 
+export function shouldUsePreviouslyCachedEntry(
+  upstreamImage: ImageUpstream,
+  previousCacheEntry: IncrementalCacheItem | undefined
+): previousCacheEntry is NonNullable<IncrementalCacheItem> & {
+  value: CachedImageValue
+} {
+  return (
+    previousCacheEntry?.value?.kind === 'IMAGE' &&
+    // If the cached image is optimized
+    previousCacheEntry.value.upstreamEtag !== previousCacheEntry.value.etag &&
+    // and the upstream etag is the same as the previous cache entry's
+    getImageEtag(upstreamImage.buffer) === previousCacheEntry.value.upstreamEtag
```

Comment:
I think you could improve performance here by using the etag from the upstream response header and fallback to the hash if its missing.

Reply:
I thought of using the upstream response header at first, but didn't like how the code turned out, or because of the [weak validation syntax](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#syntax) (it can start with `W/`, the etag value would need to be base64-encoded correctly to prevent any errors.

What I had previously was alter `ImageUpstream` to include `etag` and then get it from the response and have the hash as fallback. Didn't quite like the implementation then, but writing an ETag extractor helper function would probably make it a bit nicer and make it possible to write tests for it.

Reply:
Added this now along with tests, not 100% happy with the tests but they should still be sufficient I think.